java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -68,300 +72,333 @@
-
-    abstract void evaluateConditionalOr(BinaryExpressionTree tree);
-
-    abstract void evaluateRelationalOperator(BinaryExpressionTree tree, SymbolicRelation operator);
-
-    @CheckForNull
-    final SymbolicValue retrieveSymbolicValue(ExpressionTree tree) {
-      ExpressionTree currentTree = tree;
-      if (isSuperOrThisMemberSelect(tree)) {
-        currentTree = ((MemberSelectExpressionTree) currentTree).identifier();
-      }
-      if (currentTree.is(Tree.Kind.IDENTIFIER)) {
-        IdentifierTree identifierTree = (IdentifierTree) currentTree;
-        Symbol symbol = identifierTree.symbol();
-        if (symbol.isVariableSymbol()) {
-          return new SymbolicValue.SymbolicVariableValue((Symbol.VariableSymbol) symbol);
-        }
+    /**
+     * evaluates a condition.
+     *
+     * a condition is an expression (of type boolean), which can occur in the following places:
+     * - condition of an if statement
+     * - condition of a loop statement: do...while, for, while (but not foreach)
+     * - left operand of a logical and && and or || operators
+     * - first operand of a ternary operator
+     * conditions of switch statement are broader and not evaluated using this method.
+     *
+     * this method evaluates the given condition. if the result is either true or false,
+     * then the execution state is returned.
+     * otherwise, if the result is unknown, then the execution state is split in two branches
+     * with result TRUE and FALSE. whenever possible constraints are also set on the variables
+     * found in the condition.
+     *
+     * @param state execution state
+     * @param tree tree representing the condition
+     * @return output states
+     */
+    private List<ExecutionState> evaluateCondition(ExecutionState executionState, ExpressionTree tree) {
+      List<ExecutionState> result = new ArrayList<>();
+      ExpressionTree expressionTree = removeCastAndParenthesis(tree);
+      Tree.Kind kind = ((JavaTree) expressionTree).getKind();
+      SymbolicRelation relation = OPERATOR_TO_RELATION.get(kind);
+      if (relation != null) {
+        evaluateConditionRelationalOperator(result, executionState, (BinaryExpressionTree) expressionTree, relation);
       } else {
-        Long value = LiteralUtils.longLiteralValue(currentTree);
-        if (value != null) {
-          return new SymbolicValue.SymbolicLongValue(value);
+        switch (kind) {
+          case CONDITIONAL_AND:
+            evaluateConditionConditionalAnd(result, executionState, (BinaryExpressionTree) expressionTree);
+            break;
+          case CONDITIONAL_OR:
+            evaluateConditionConditionalOr(result, executionState, (BinaryExpressionTree) expressionTree);
+            break;
+          case IDENTIFIER:
+            evaluateConditionIdentifier(result, executionState, ((IdentifierTree) expressionTree).symbol());
+            break;
+          default:
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1151)